GH-8757 let ObjectStringMessageConverter.toMessage() accept arbitrary payload

mrpiggi opened the pull request:
Fixes #8757

* enhance ObjectStringMessageConverter to process arbitrary payload
* add appropriate tests

mrpiggi committed:
add ctor ObjectStringMessageConverter(Charset)

* create tests for ObjectStringMessageConverter

mrpiggi commented:
As https://github.com/spring-projects/spring-integration/blob/22cbe14586a29aafc4d444c12a80c881ed8764d8/spring-integration-core/src/main/java/org/springframework/integration/support/converter/ObjectStringMessageConverter.java#L55-L59
is now not guarded by
https://github.com/spring-projects/spring-integration/blob/22cbe14586a29aafc4d444c12a80c881ed8764d8/spring-integration-core/src/main/java/org/springframework/integration/support/converter/ObjectStringMessageConverter.java#L50-L53
anymore, it is debatable, how `Class<?> targetClass` should be treated. It could either be enforced to `String.class` (currently implemented) or be unconditionally ignored or something in between like `Assert.isTrue(targetClass == null || targetClass == String.class)`

mrpiggi commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/ObjectStringMessageConverter.java:
```diff
@@ -43,4 +65,10 @@ protected Object convertFromInternal(Message<?> message, Class<?> targetClass, O
+	@Nullable
+	protected Object convertToInternal(
+			Object payload, @Nullable MessageHeaders headers, @Nullable Object conversionHint) {
+		return super.convertToInternal(payload.toString(), headers, conversionHint);
```
I already suspected that, as `ObjectStringMessageConverter` appears a bit odd. Maybe another name like `FallbackStringMessageConverter` would have been a better name.

mrpiggi commented:
As this boils down to adding only ctor `ObjectStringMessageConverter(Charset defaultCharset)` should I close this PR and create a new one with a sinble commit?

artembilan commented:
Bo need in a fresh PR: we squash commits on merge . Yes, you can add nullability for this package just in this PR.

mrpiggi commented:
squashed current state as I had to rebase on upstream anyways. will add `@NonNullApi ` on this PR this evening.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/ObjectStringMessageConverter.java:
```diff
@@ -24,14 +26,30 @@
+ * <p>
+ * This class meant to be used as a fallback converter internally when deserializing messages. Therefore, only
+ * {@link org.springframework.messaging.converter.AbstractMessageConverter#fromMessage(Message, Class) fromMessage}
+ * method should be called with {@code String.class} as {@code targetClass}, obviously.To be explicit, using method
```
Sorry, I don't find words like `obviously` and `To be explicit` as formal technical documentation language.
I also don't this that we need to mention `toMessage()` in this Javadoc at all.
It is really the superclass responsibility to express its nature.
we override only `convertFromInternal()` and we should explain for what reason.
Nothing more to make the reader to feel silly.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/support/converter/ObjectStringMessageConverterTests.java:
```diff
@@ -0,0 +1,69 @@
+ *
+ */
+class ObjectStringMessageConverterTests {
+	private static final ArbitraryPayload PAYLOAD = new ArbitraryPayload(1234, "ÄÖÜ");
```
We prefer to wrap every class member into blank lines, even the last method (or inner class, record) before final `}` should have a blank line after its declaration.

This way the code is much readable, especially during review here on GH.

mrpiggi committed:
fix codestyle issues

mrpiggi committed:
revise JavaDoc

mrpiggi committed:
@NonNullApi for org.springframework.integration.support.converter

- simple cases first

mrpiggi committed:
revise SimpleMessageConverter

- inboundMessageMapper and outboundMessageMapper are initialized to their default implementations
- both are set in non-default ctors with final setters

mrpiggi commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/ConfigurableCompositeMessageConverter.java:
```diff
@@ -97,6 +99,7 @@ public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
+			// TODO how to deal with this.beanFactory == null? just use DefaultConversionService.getSharedInstance()?
 			ConversionService conversionService = IntegrationUtils.getConversionService(this.beanFactory);
 			if (conversionService == null) {
 				conversionService = DefaultConversionService.getSharedInstance();
```
If we consider `this.beanFactory` as `@Nullable` should we set `conversionService = DefaultConversionService.getSharedInstance();` as well or is some other behavior required?

artembilan replied:
The `BeanFactoryAware.setBeanFactory(BeanFactory)` is not nullable.
Not sure why do yo u make such an assumption over here then...

mrpiggi replied:
Well, the field `private BeanFactory beanFactory` is checked for `null` in several places throughout the whole framework although it is annotated with `@Nullable` only at very few places. I am fine with removing the annotation although it seems inconsistent regarding `@NonNullFields`

mrpiggi replied:
With inconsistency I mean that  field `private BeanFactory beanFactory` as actually marked as `@NonNull` via `@NonNullFields` but actually can be `null` until `setBeanFactory` was invoked. But as I said, this holds for the major part of the framework.

mrpiggi replied:
Actually, for `ConfigurableCompositeMessageConverter` I did not change anything. Currently, there is no check `this.beanFactory != null` in `main`.

I just removed the check from `DefaultDatatypeChannelMessageConverter.setBeanFactory()` as you told me, it must not be called with `null` as an argument. In this case, getting a NPE is the expected behavior, isn't it?

artembilan replied:
That's correct, but since we talk about Spring, it is expected that these classes have to be registered as beans in the application context.
Therefore that `setBeanFactory()` is always called and really never with `null`.
So, we may assume that it is null by some end-user mistake or we can just bite a bullet and let it fly for time being!

mrpiggi replied:
I am totally aware of that. `@NonNullFields` is more or less a tool for developers and actually no part of the public API. I am fine with every call you make.

mrpiggi commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/MapMessageConverter.java:
```diff
@@ -43,10 +43,12 @@
 
-	private volatile String[] headerNames;
+	// TODO is initializing with an empty array correct or should toMessage() return 'null' if never set?
+	private volatile String[] headerNames = {};
```
This is probably not desired. Should I add something like `Assert.notEmpty(this.headerNames, "at least one header name must be set with setHeaderNames()");` for `toMessage()` and `fromMessage()`

artembilan replied:
I think we need to make it nullable and implement respective logic.
It is going to be more efficient to avoid some extra calls if this one is just null.

mrpiggi replied:
The question still remains, what logic should be implemented. Should `this.headerNames == null` be just handled like an empty array or is a different behavior needed?

artembilan replied:
I think what you did is enough.
It really looked like a bug by default where this `headerNames` is null and it is not handled properly.
Now it is OK. 
Let's remove that TODO!
Thank you!

mrpiggi replied:
Fine. I adapted `setHeaderNames(String... headerNames)` to use an empty array if called with `null` or an empty array

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/DefaultDatatypeChannelMessageConverter.java:
```diff
@@ -58,7 +59,7 @@ public void setConversionService(ConversionService conversionService) {
 
 	@Override
-	public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
+	public void setBeanFactory(@Nullable BeanFactory beanFactory) throws BeansException {
```
Same here: we need to rework the logic with an assumption that it is not null.

mrpiggi replied:
well in this case, we could just drop the check in the following if statement

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/MapMessageConverter.java:
```diff
@@ -43,10 +43,12 @@
 
 	private volatile boolean filterHeadersInToMessage;
 
+	@Nullable
```
Same here: it is not supposed to be `null`

artembilan reviewed (changes requested):
I will look into failure of the MongoDB module myself.

mrpiggi committed:
remove `@Nullable` from `private BeanFactory beanFactory`

mrpiggi committed:
setBeanFactory() can not be called with `null`

mrpiggi committed:
handle empty `String... headerNames`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/MapMessageConverter.java:
```diff
@@ -77,12 +78,16 @@ protected MessageBuilderFactory getMessageBuilderFactory() {
-		Assert.notEmpty(headerNames, "at least one header name is required");
-		this.headerNames = Arrays.copyOf(headerNames, headerNames.length);
+		if (ObjectUtils.isEmpty(headerNames)) {
+			this.headerNames = new String[] {};
```
Why? With `Assert.notEmpty()` we just disallow to set a null into this property.
I don't see anything what really has to be fixed in this method.
The default one is empty array, but if it is set, it must not be empty.
That's it. What wrong do you see with that logic?

mrpiggi replied:
With the current implementation, it can be changed to anything but the default state during runtime. So setting back to an empty array is not possible.

artembilan replied:
OK. That makes sense.
I'll pull your PR locally shortly for final review and possible merge.

Thank you!

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/29186e2b4609d30342c19e467b11e9eb45b1324d

@mrpiggi ,

thank you for contribution; looking forward for more!
